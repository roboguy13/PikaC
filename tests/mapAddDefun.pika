%specialize map plus

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

map : (Int -> Int) -> Sll -> Sll;
map f Nil := Nil [Sll];
map f (Cons a as) := Cons (f a) (map f as) [Sll];

-- mapAdd1 : Sll -> Sll;
-- mapAdd1 xs := map (\x. x + 1) xs;

-- %test "mapAdd1" Sll:
--   mapAdd 1 (Cons 10 (Cons 20 (Cons 30 (Nil [Sll]) [Sll]) [Sll]) [Sll]);
