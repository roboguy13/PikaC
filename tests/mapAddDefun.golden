void map__lambda_(loc x, loc x1);
void map__lambda_(loc x, loc x1) {
 if (x == 0) {
  WRITE_LOC(x1, 0, 0);
 } else {
  loc h = READ_LOC(x, 0);
  loc n = READ_LOC(x, 1);
  map__lambda_(n, x1);
  loc x3 = READ_LOC(x1, 0);
  loc x = (loc)malloc(2 * sizeof(loc));
  WRITE_LOC(x1, 0, x);
  WRITE_LOC(x, 1, x3);
  WRITE_LOC(x, 0, (long)h + (long)1);
 }
}